### PR TITLE
Loofah 34 Get skills by craftIds

### PR DESCRIPTION
**Pull request by elishkaflint**

### CODE REVIEW

File: src/test/java/com/loofah/graph/api/services/SkillServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        Skill expectedSkill = Skill.builder().build();
+        when(skillRepository.findById("id")).thenReturn(Optional.of(expectedSkill));
+
+        Skill actualSkill = skillService.getById("id");
+
+        assertEquals(expectedSkill, actualSkill);
+    }
+
+    @Test
+    public void getWithFilter_categoryId() {
```
Comment (elishkaflint) [inline src/test/java/com/loofah/graph/api/services/SkillServiceTest.java]: i don't love this test, if anyone has any ideas to test the logic to build the query?

**Comment by elishkaflint** [comment]
i've added some new services on this PR as I think queries are starting to hold a bit more business logic (and this will probably increase in future) - let me know what you think of this structure!

File: src/test/java/com/loofah/graph/api/services/SkillServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+        Query expectedQuery = new Query();
+        expectedQuery.addCriteria(Criteria.where("gradeId").is(GRADE_ID));
+
+        skillService.getWithFilter(skillFilter);
+        verify(skillRepository).findWithQuery(expectedQuery, Skill.class);
+    }
+
+    @Test
+    public void getWithFilter_gradeId() {
```
Comment (vijamesDD) [inline src/test/java/com/loofah/graph/api/services/SkillServiceTest.java]: Did you intend to leave these empty?

**Review by vijamesDD — APPROVED** [review]
💯  Loving these changes. I love that this repo is so fluid - constantly updating and tweaking things here and there to improve it 👏

File: src/main/java/com/loofah/graph/api/services/SkillService.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        optionalFilter.map(SkillFilter::getCategoryId).ifPresent(
+                categoryId -> query.addCriteria(Criteria.where("categoryId").is(categoryId))
+        );
+
+        optionalFilter.map(SkillFilter::getGradeId).ifPresent(
+                gradeId -> query.addCriteria(Criteria.where("gradeId").is(gradeId))
+        );
+
+        optionalFilter.map(SkillFilter::getCraftIds).ifPresent(
+                craftIds -> query.addCriteria(Criteria.where("craftIds").in(craftIds))
```
Comment (Learner5200) [inline src/main/java/com/loofah/graph/api/services/SkillService.java:49]: If I’m following this logic correctly, it means we’ll match only skills for which their full set of associated  crafts are included in our search filter. So if we searched for backend skills, we would get all the backend only skills, but we wouldn’t get skills that are associated with both backend and devops. Is this the desired behaviour?

File: src/test/java/com/loofah/graph/api/services/SkillServiceTest.java (hunk 1 above)
Comment (Learner5200) [inline src/test/java/com/loofah/graph/api/services/SkillServiceTest.java]: The body of this test is testing grade ID filtering, not category.

File: src/test/java/com/loofah/graph/api/services/SkillServiceTest.java (hunk 2 above)
Comment (Learner5200) [reply src/test/java/com/loofah/graph/api/services/SkillServiceTest.java]: +1 - I also think it would be good to have unit tests here for all the filters (grade/category/craft), and for no filters (or a null filter object).

Comment (elishkaflint) [reply src/test/java/com/loofah/graph/api/services/SkillServiceTest.java]: yeah i left them empty and i'm a bit unsure whether the test from line 41 makes sense? it seems bit atypical to create an "expected parameter" and verify that it's used. i also think it might not work when the query has an array when it's to do with crafts.

definitely want to make sure the scenarios i've left blank are tested but does anyone have any thoughts on how to write a better test?

File: src/main/java/com/loofah/graph/api/services/SkillService.java [hunk 4]
```diff
@@ -2,55 +2,34 @@
...
+        this.mongoQueryProvider = mongoQueryProvider;
     }
 
     public Skill getById(String id) {
         return skillRepository.findById(id).get();
     }
 
     public List<Skill> getWithFilter(SkillFilter skillFilter) {
-        Query query = buildQuery(skillFilter);
+        Query query = mongoQueryProvider.buildMongoQuery(skillFilter);
```
Comment (Learner5200) [inline src/main/java/com/loofah/graph/api/services/SkillService.java:31]: Nitpick: this service class still has a dependency on our database implementation (i.e. it's depending on `Query` and`MongoQueryProvider` classes, and on a `SkillRepository` interface which extends a mongo interface). This means that if we ever do want to change the database implementation, this service will have to change too.

Well beyond the scope of this ticket to fix this, but something for us to bear in mind.

**Review by Learner5200 — APPROVED** [review]

File: src/main/java/com/loofah/graph/api/services/SkillService.java (hunk 4 above)
Comment (elishkaflint) [reply src/main/java/com/loofah/graph/api/services/SkillService.java:31]: As discussed, great shout and will bring through another ticket immediately after this one to improve the design a bit here
